addon/v1alpha1/types_managedclusteraddon.go

```diff
@@ -29,8 +29,68 @@ type ManagedClusterAddOn struct {
+// an addon agent on managed cluster.
 type ManagedClusterAddOnSpec struct {
+	// installNamespace is the namespace on the managed cluster to install the addon agent.
+	// If it is not set, open-cluster-management-addon namespace is used to install the addon agent.
+	// +optional
+	InstallNamespace string `json:"installNamespace,omitempty"`
+
+	// leaseDurationSeconds is used to coordinate the lease update time of addon agents on the managed cluster.
+	// The addon agent should maintain a lease with the name of addon name in the namespace of installNamespace.
+	// The Klusterlet agent will monitor the lease and update the status of the related managedclusteraddon
+	// resource accordingly.  If its value is zero, the addon agent should update its lease every 60 seconds by default
+	// +optional
+	LeaseDurationSeconds int32 `json:"leaseDurationSeconds,omitempty"`
+
+	// registrations defines the conifigurations for the addon agent to register to hub. With the registration defined,
+	// The addon agent can access to kube apiserver with kube style API or other endpoints on hub cluster with client
+	// certificate authentication. A csr will be created per registration configuration. If more than one
+	// registrationConfig is defined, a csr will be created for each registration configuration. It is not allowed that
+	// multiple registrationConfigs have the same signer name. After the csr is approved on the hub cluster, the klusterlet
+	// agent will create a secret in the installNamespace for the registrationConfig. If the signerName is
+	// "kubernetes.io/kube-apiserver-client", the secret name will be "{addon name}-hub-kubeconfig" whose contents includes
+	// key/cert and kubeconfig. Otherwise, the secret name will be "{addon name}-{signer name}-client-cert" whose contents includes key/cert.
+	// +optional
+	Registrations []RegistrationConfig `json:"registrations,omitempty"`
+}
+
+// RegistrationConfig defines the configuration of the addon agent to register to hub. The Klusterlet agent will
+// create a csr for the addon agent with the registrationConfig.
+type RegistrationConfig struct {
+	// signerName is the name of signer that addon agent will use to create csr.
+	// +required
+	SignerName string `json:"signerName"`
+
+	// subject is the user subject of the addon agent to be registered to the hub.
+	// If it is not set, the addon agent will have the default subject
+	// "subject": {
+	//	"user": "system:open-cluster-management:addon:{addonName}:{clusterName}:{agentName}",
+	//  "groups: ["system:open-cluster-management:addon", "system:open-cluster-management:addon:{addonName}"]
+	// }
+	//
+	// +optional
+	Subject Subject `json:"subject,omitempty"`
+}
+
+// Subject is the user subject of the addon agent to be registered to the hub.
+type Subject struct {
+	// user is the user name of the addon agent.
+	User string `json:"user"`
+
+	// uid is the user id of the addon agent.
+	// +optional
+	UID string `json:"uid,omitempty"`
```

Comment:
UID looks odd to me.  I don't think I'd let a client set this value.

Reply:
removed